### DISCUSSION TITLE
using await in a setstate function

### DISCUSSION THREAD

**Discussion by PazminoJose** [pmndrs/zustand#1935] [Q&A] [answered]
Hello, I have a question.
I create an store like this with a prop clearFilter that set my reportFilter and dateFilter to the initial state
```js
export interface ReportState {
    records: UserCompletedRoutineReport[],
    setRecords: (records: UserCompletedRoutineReport[]) => void,
    getRecords: (query?: FilterQuery) => Promise<void>;
    reportFilter: ReportFilter;
    setReportFilter: (reportFilter: ReportFilter) => void;
    dateFilter: DateFilter;
    setDateFilter: (dateFilter: DateFilter) => void;
    clearFilter: () => void;
}

const initialState: ReportState = {
    records: [],
    setRecords: () => { },
    getRecords: async () => { },
    dateFilter: {},
    setDateFilter: () => { },
    reportFilter: {},
    setReportFilter: () => { },
    clearFilter: async () => { }
}

export const useReportStore = create<ReportState>((set) => ({
    records: initialState.records,
    setRecords: (records) => { set({ records }) },
    getRecords: async (query) => {
        const res = await getReportsService(query)
        set({ records: res })
    },
    dateFilter: initialState.dateFilter,
    setDateFilter: (dateFilter: DateFilter) => set((state) => ({ dateFilter: { ...state.dateFilter, ...dateFilter } })),
    reportFilter: initialState.reportFilter,
    setReportFilter: (reportFilter: ReportFilter) => set((state) => ({ reportFilter: { ...state.reportFilter, ...reportFilter } })),
    clearFilter: () => {
        set({ dateFilter: initialState.dateFilter, reportFilter: initialState.reportFilter });
    },
}));
```
in other hand i use it with react query with a custom hook that have an query as parameter
```js
export const useGetRecordsQuery = ({ query, onSuccess, enabled }: { query?: FilterQuery, onSuccess?: (data: UserCompletedRoutineReport[]) => void, enabled?: boolean }) => {
  return useQuery({
    queryKey: [QueryKey.REPORTS],
    queryFn: () => getReportsService(query),
    onSuccess,
    enabled
  });
};
  const { refetch } = useGetRecordsQuery({
        query: { ...reportFilter, ...dateFilter },
        onSuccess: (data) => {
            setRecords(data)
        },
        enabled: false
    });
```
when i call a function handleClearFilter i first call my clearFilter from store and then a do refetch but with this my refetch takes the old values from reportFilter and dateFilter and don´t fetch the correct data because the query param keeps the old values.

```js
  const handleClearFilter = async () => {
        clearFilter();
        await refetch();
    }
```
but if i add an await to the clearFilter the query takes the new values and do the refetch with teh new data

```js
  const handleClearFilter = async () => {
        await clearFilter();
        await refetch();
    }
```
This work for me but idk why and i don´t really know if this is ok because vs code say that the await don´t have any effect in my clearFilter

**Comment by dai-shi** [comment]
How does this do?
```js
const handleClearFilter = async () => {
  clearFilter();
  await Promise.resolve();
  await refetch();
};
```

**Reply by PazminoJose** [reply]
It does the same as 
```js
 const handleClearFilter = async () => {
        await clearFilter();
        await refetch();
    }
```

**Reply by dai-shi** [reply]
Which means, it's not the issue of clearFilter, but refetch wants to wait a little bit before calling.

**Reply by PazminoJose** [reply]
I see, so just to know, it is ok to do the Promise.resolve() in terms of good practice to let the refetch wait what it needs?

**Reply by dai-shi** [reply]
Ah, I think I understand what's happening. You are using zustand state through hooks for your query. That's why it's delayed.
I don't think you need `refetch` if you make query id to include those filters.

**Reply by PazminoJose** [reply]
Yes i use zutand  state for my queries
something like this
```js
  const {
        dateFilter,
        reportFilter,
        setDateFilter,
        setReportFilter,
        clearFilter,
        setRecords
    } = useReportStore((state) => ({
        dateFilter: state.dateFilter,
        reportFilter: state.reportFilter,
        setDateFilter: state.setDateFilter,
        setReportFilter: state.setReportFilter,
        clearFilter: state.clearFilter,
        setRecords: state.setRecords
    }), shallow);

    const { refetch } = useGetRecordsQuery({
        query: { ...reportFilter, ...dateFilter },
        onSuccess: (data) => {
            setRecords(data)
        },
        enabled: false
    });
```
I need to use the refetch because i have two buttons, one to filter after select all the filter that the user wants, i do this because if i include the filter in the query key it will refetch every time i change a filter and for me it´s a lot of request every time so for this i include a button an let the user decide when he wants to filter the data. and my second buttons is for clear all the filters and i would like to refetch the data after clean all the filters so the user can se the change.

**Reply by dai-shi** [reply, marked as answer]
I don't think it's Zustand's issue specifically, and you can reproduce the same behavior with `useState` instead.
In such a case, you might want to ask for help from RQ community, because they should have some best practice.

**Reply by dbritto-dev** [reply]
@BlueBird0399 you shouldn't use `onSuccess` to set up a state, you can read more about `onSuccess` and `onError` here: https://twitter.com/TkDodo/status/1434568298138570760?t=Ct-Mju6ju4BgYrexPcu5WQ&s=19

BTW, they're planning to remove them on `v5`

**Reply by PazminoJose** [reply]
Thanks for the advice and thanks for your time, I´m gonna look more into RQ I just start using the package.
